# FateMap_Goyal2023

Public Repository containing all code used in the manuscript Goyal et al. Biorxiv link: https://www.biorxiv.org/content/10.1101/2021.12.08.471833v1.full

The code is organized into two folders:

1. extractionScripts: This folder contains scripts that were used to extract data from raw files. The folder contains subfolders, each of which correspond to different types of datasets (e.g. 10X, imaging, Whole genome sequencing).
2. plotScripts: This folder contains scripts that work on the extracted data from #1 to make the plots that went into the paper.

All relevant data used have respective accession numbers and DOIs. See below:

All raw and processed imaging data generated in this study can be found on BioStudies (Accession S-BIAD696). All raw and processed FateMap single-cell barcoding data generated in this study can be found at Figshare (DOI 1: 10.6084/m9.figshare.22798952 ; DOI 2: 10.6084/m9.figshare.22802888). All raw and processed gDNA barcoding data generated in this study can be found at Figshare (DOI: 10.6084/m9.figshare.22806494). All raw and processed whole genome sequencing data of individual clones used in this manuscript can be found at BioProject Accession PRJNA972638 and figshare (10.6084/m9.figshare.23255273), respectively. All raw and processed bulk RNA Sequencing data used in this manuscript can be found at GEO Accession GSE233622. All raw and processed single-cell RNA Sequencing data used in this manuscript can be found at GEO Accession: GSE233766. All raw and processed data from the GeoMx spatial transcriptomics used in this manuscript can be found at BioProject Accession PRJNA976929 and Figshare (10.6084/m9.figshare.23248199), respectively.

If you have any questions about the code and data, please address correspondence to Arjun Raj (arjunrajlab@gmail.com) and Yogesh Goyal (yogesh.goyal@northwestern.edu).
